### DISCUSSION TITLE
Quarkus Keycloak client authentication using x509

### DISCUSSION THREAD

**Discussion by tlann** [keycloak/keycloak#15580] [Archive]
I have been trying to figure how to get client authentication working using x509 certificates in the Quarkus version of Keycloak.  @dasniko has a helpful [video](https://www.youtube.com/watch?v=yq1hzNs1JQU&t=1086s) of doing it with the legacy version of Keycloak.  I tried to follow what he and the [Keycloak server admin guide suggests](https://www.keycloak.org/docs/latest/server_admin/#_x509).  

The expected behavior, based on @dasniko 's video is a certificate selection window should open up upon visiting the login url.  However,  only the login screen is displayed.  

I have created a [simple example](https://github.com/tlann/keycloak-mtls) in github to demonstrate what I'm attempting.  There is a script to generate the certificates and import them into a truststore and keystore jks.  The script then copies the jks files to a directory.  The compose.yml file mounts the cert directory when starting Keycloak up and points Keycloak to them as a keystore and truststore.  Their is also a realm export to configure things.  
I have seen reports others have been able to do this.  Do any of the symptoms sound familiar?  Any suggestions on troubleshooting this?

**Comment by thomasdarimont** [comment]
Hi @tlann did you mange to solve this issue?

**Reply by thomasdarimont** [reply]
Just got it working with Keycloak 20.0.2:

I used the following options:
```
      - "--https-client-auth=request"
      - "--https-trust-store-file=/opt/keycloak/conf/truststore.p12"
      - "--https-trust-store-password=changeit"
      - "--https-protocols=TLSv1.2,TLSv1.3"
```

The `truststore.p12` had to contain the rootCA certificate used to sign the client certs ... or the explicit list of client certs that you want to trust explicitly.

Without this, the `org.keycloak.authentication.authenticators.x509.AbstractX509ClientCertificateAuthenticator#getCertificateChain` always returned null, which leads to the abortion of the X509 authenticator.

The authentication flow looks like this:
![image](https://user-images.githubusercontent.com/314690/207745227-928150cf-c8f4-4f8b-b623-933bdcc8a445.png)

The X509 Validate Username Form Authenticator config looks like this:
![image](https://user-images.githubusercontent.com/314690/207745321-dbfb41e4-6f37-4db5-ab3c-6b0ca5214eae.png)

Login with Cert selection and confirmation
![keycloakx-x509-confirm-Peek 2022-12-15 01-41](https://user-images.githubusercontent.com/314690/207745832-76c56ec9-5682-4e04-a0cf-cb3bb586441d.gif)

Login with Cert selection and confirmation bypass
![keycloakx-x509-bypass-confirm-Peek 2022-12-15 01-41](https://user-images.githubusercontent.com/314690/207745899-064697a5-5e55-4865-b274-1dd55f2b57de.gif)

**Reply by tlann** [reply]
Thank you for checking this out.  The differences I see in my example that sticks out to me are https-client-auth=request and my example had require.  I also didn't have https-protocols set.  
We also were using a jks cert store vs a p12 one.  
Thanks again

**Comment by bikebaba** [comment]
You said this was Keycloak 20.0.2. I'm using the same, but my GUI looks different and I cannot add the x509 flows that I did back in version 15. Anything else I need to do? Any idea?

**Reply by thomasdarimont** [reply]
I used the old admin-console theme in my example.

**Comment by tlann** [comment]
@thomasdarimont I was wondering if you had a little more time to help.  Or anyone else available.  I would be willing to pay for some time to help with this issue.  Whether is be a working example or something else.  

As before,  the code I am trying to run is [here](https://github.com/tlann/keycloak-mtls/tree/main)
I finally got time to work on this again and I'm still having the problem where the cert selection window doesn't pop up when I get to that page.  

The  https_client_auth parameter is set to request as shown.  But I used an environment variable to do that as opposed to a command line parameter.  

The rootCA cert is also added to the truststore as seen [here](https://github.com/tlann/keycloak-mtls/blob/main/certs/x509.sh#L43).  
```
  services:
    keycloak:
      image: quay.io/keycloak/keycloak:20.0.2
      command: [ "start-dev" ]
      environment:
        JAVA_OPTS_APPEND: "-Dcom.redhat.fips=false"
        KC_HOSTNAME_STRICT: "false"
        KEYCLOAK_ADMIN: admin
        KEYCLOAK_ADMIN_PASSWORD: admin
        KC_LOG_LEVEL: info,org.keycloak.authentication:trace
        KC_HTTPS_KEY_STORE_FILE: /genCerts/keystore.jks
        KC_HTTPS_KEY_STORE_PASSWORD: changeit
        KC_HTTPS_TRUST_STORE_FILE: /genCerts/truststore.jks
        KC_SPI_TRUSTSTORE_FILE_FILE: /genCerts/truststore.jks
        KC_SPI_TRUSTSTORE_FILE_PASSWORD: changeit
        KC_HTTPS_TRUST_STORE_PASSWORD: changeit
        KC_SPI_TRUSTSTORE_FILE_HOSTNAME_VERIFICATION_POLICY: ANY
        KC_HTTPS_CLIENT_AUTH: request
        KC_HTTPS_PROTOCOLS: TLSv1.2,TLSv1.3
        KC_FEATURES_DISABLED: admin2
        KC_FEATURES: admin
        KC_SPI_IMPORT_SINGLE_FILE_FILE: x509-realm-export.json
      ports:
        - 8080:8080
        - 8443:8443
      volumes:
        - ./genCerts:/genCerts
        - ./realm:/opt/keycloak/data/import
```

The x509 browser flow looks like the following 
![image](https://github.com/keycloak/keycloak/assets/13425578/9ee0def2-c569-432c-9a0b-2bec25bf0786)

The x509-config Validate Username Form Authenticator config looks like this: looks like 
![image](https://github.com/keycloak/keycloak/assets/13425578/7abd73c6-be29-4208-b163-ffbb6b4c50a1)

I turned on debugging and get the following log messages when going to the login [page](https://myHost:8443/realms/x509/protocol/openid-connect/auth?client_id=account-console&redirect_uri=https%3A%2F%2Fvdi-rh-21%3A8443%2Frealms%2Fx509%2Faccount%2F%23%2F&state=90126495-6c91-4fcb-9de8-309d581d71c4&response_mode=fragment&response_type=code&scope=openid&nonce=fa99d921-5dba-41f1-82ae-07d734e125c9&code_challenge=bc5HF-XQ6P0t7ZHXiSwPF0iU585eRDODoqNoyQmLxPM&code_challenge_method=S256) .  But not when I go to the account page at https://myHost:8443/realms/x509/account/

I have turned logging on for org.keycloak.authentication and I get the following messages only when I got to the username password page.  
```
mtls-2-keycloak-1  | 2023-06-12 21:36:19,730 DEBUG [org.keycloak.authentication.AuthenticationProcessor] (executor-thread-215) AUTHENTICATE
mtls-2-keycloak-1  | 2023-06-12 21:36:19,731 DEBUG [org.keycloak.authentication.AuthenticationProcessor] (executor-thread-215) AUTHENTICATE ONLY
mtls-2-keycloak-1  | 2023-06-12 21:36:19,731 DEBUG [org.keycloak.authentication.DefaultAuthenticationFlow] (executor-thread-215) processFlow: browser x509
mtls-2-keycloak-1  | 2023-06-12 21:36:19,731 DEBUG [org.keycloak.authentication.DefaultAuthenticationFlow] (executor-thread-215) check execution: 'auth-cookie', requirement: 'ALTERNATIVE'
mtls-2-keycloak-1  | 2023-06-12 21:36:19,731 DEBUG [org.keycloak.authentication.DefaultAuthenticationFlow] (executor-thread-215) authenticator: auth-cookie
mtls-2-keycloak-1  | 2023-06-12 21:36:19,731 DEBUG [org.keycloak.authentication.AuthenticationSelectionResolver] (executor-thread-215) Going through the flow 'browser x509' for adding executions
mtls-2-keycloak-1  | 2023-06-12 21:36:19,732 DEBUG [org.keycloak.authentication.AuthenticationSelectionResolver] (executor-thread-215) Going through the flow 'browser x509 forms' for adding executions
mtls-2-keycloak-1  | 2023-06-12 21:36:19,732 DEBUG [org.keycloak.authentication.AuthenticationSelectionResolver] (executor-thread-215) Selections when trying execution 'auth-cookie' : [ authSelection - auth-cookie,  authSelection - auth-x509-client-username-form,  authSelection - auth-username-password-form]
mtls-2-keycloak-1  | 2023-06-12 21:36:19,732 DEBUG [org.keycloak.authentication.DefaultAuthenticationFlow] (executor-thread-215) invoke authenticator.authenticate: auth-cookie
mtls-2-keycloak-1  | 2023-06-12 21:36:19,732 DEBUG [org.keycloak.authentication.DefaultAuthenticationFlow] (executor-thread-215) authenticator ATTEMPTED: auth-cookie
mtls-2-keycloak-1  | 2023-06-12 21:36:19,732 TRACE [org.keycloak.authentication.DefaultAuthenticationFlow] (executor-thread-215) Set execution status: Execution: auth-cookie, status: ATTEMPTED
mtls-2-keycloak-1  | 2023-06-12 21:36:19,732 DEBUG [org.keycloak.authentication.DefaultAuthenticationFlow] (executor-thread-215) check execution: 'auth-x509-client-username-form', requirement: 'ALTERNATIVE'
mtls-2-keycloak-1  | 2023-06-12 21:36:19,732 DEBUG [org.keycloak.authentication.DefaultAuthenticationFlow] (executor-thread-215) authenticator: auth-x509-client-username-form
mtls-2-keycloak-1  | 2023-06-12 21:36:19,732 DEBUG [org.keycloak.authentication.AuthenticationSelectionResolver] (executor-thread-215) Going through the flow 'browser x509' for adding executions
mtls-2-keycloak-1  | 2023-06-12 21:36:19,732 DEBUG [org.keycloak.authentication.AuthenticationSelectionResolver] (executor-thread-215) Going through the flow 'browser x509 forms' for adding executions
mtls-2-keycloak-1  | 2023-06-12 21:36:19,732 DEBUG [org.keycloak.authentication.AuthenticationSelectionResolver] (executor-thread-215) Selections when trying execution 'auth-x509-client-username-form' : [ authSelection - auth-x509-client-username-form,  authSelection - auth-username-password-form]
mtls-2-keycloak-1  | 2023-06-12 21:36:19,733 DEBUG [org.keycloak.authentication.DefaultAuthenticationFlow] (executor-thread-215) invoke authenticator.authenticate: auth-x509-client-username-form
mtls-2-keycloak-1  | 2023-06-12 21:36:19,735 DEBUG [org.keycloak.authentication.DefaultAuthenticationFlow] (executor-thread-215) authenticator ATTEMPTED: auth-x509-client-username-form
mtls-2-keycloak-1  | 2023-06-12 21:36:19,736 TRACE [org.keycloak.authentication.DefaultAuthenticationFlow] (executor-thread-215) Set execution status: Execution: auth-x509-client-username-form, status: ATTEMPTED
mtls-2-keycloak-1  | 2023-06-12 21:36:19,736 DEBUG [org.keycloak.authentication.DefaultAuthenticationFlow] (executor-thread-215) check execution: 'browser x509 forms flow', requirement: 'ALTERNATIVE'
mtls-2-keycloak-1  | 2023-06-12 21:36:19,736 DEBUG [org.keycloak.authentication.DefaultAuthenticationFlow] (executor-thread-215) processFlow: browser x509 forms
mtls-2-keycloak-1  | 2023-06-12 21:36:19,736 DEBUG [org.keycloak.authentication.DefaultAuthenticationFlow] (executor-thread-215) check execution: 'auth-username-password-form', requirement: 'REQUIRED'
mtls-2-keycloak-1  | 2023-06-12 21:36:19,736 DEBUG [org.keycloak.authentication.DefaultAuthenticationFlow] (executor-thread-215) authenticator: auth-username-password-form
mtls-2-keycloak-1  | 2023-06-12 21:36:19,737 DEBUG [org.keycloak.authentication.AuthenticationSelectionResolver] (executor-thread-215) Going through the flow 'browser x509' for adding executions
mtls-2-keycloak-1  | 2023-06-12 21:36:19,737 DEBUG [org.keycloak.authentication.AuthenticationSelectionResolver] (executor-thread-215) Going through the flow 'browser x509 forms' for adding executions
mtls-2-keycloak-1  | 2023-06-12 21:36:19,737 DEBUG [org.keycloak.authentication.AuthenticationSelectionResolver] (executor-thread-215) Selections when trying execution 'auth-username-password-form' : [ authSelection - auth-username-password-form]
mtls-2-keycloak-1  | 2023-06-12 21:36:19,737 DEBUG [org.keycloak.authentication.DefaultAuthenticationFlow] (executor-thread-215) invoke authenticator.authenticate: auth-username-password-form
mtls-2-keycloak-1  | 2023-06-12 21:36:19,753 TRACE [org.keycloak.authentication.DefaultAuthenticationFlow] (executor-thread-215) Set execution status: Execution: auth-username-password-form, status: CHALLENGED
mtls-2-keycloak-1  | 2023-06-12 21:36:19,753 TRACE [org.keycloak.authentication.DefaultAuthenticationFlow] (executor-thread-215) Set execution status: Execution: browser x509 forms flow, status: CHALLENGED
```

**Reply by tlann** [reply]
I figured out my issue.  Certs weren't being imported properly to the keystore vs the truststore.  I'm going to update my example in github so that anyone else that comes across the issue can have a working example.
